src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -1036,6 +1099,74 @@ private static bool IsPrimaryKeySort(SearchOptions searchOptions)
             return searchOptions.Sort.All(s => s.searchParameterInfo.Name is SearchParameterNames.ResourceType or SearchParameterNames.LastUpdated);
         }
 
+        private bool IsSortValueNeeded(SearchOptions context)
+        {
+            if (context.Sort.Count == 0)
+            {
+                return false;
+            }
+
+            if (IsPrimaryKeySort(context))
+            {
+                return false;
+            }
+
+            foreach (var searchParamTableExpression in _rootExpression.SearchParamTableExpressions)
+            {
+                if (searchParamTableExpression.Kind == SearchParamTableExpressionKind.Sort ||
+                    searchParamTableExpression.Kind == SearchParamTableExpressionKind.SortWithFilter)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static void GetSortRelatedDetails(
+            SearchOptions context,
+            out SortOrder sortOrder,
+            out ContinuationToken continuationToken,
+            out object sortValue,
+            out Column sortColumnName)
```

Comment:
This looks like it should just return an internal Class i.e. SortContext?

Reply:
Added an internal class `SortContext` to keep track of these properties